src/client/executor.rs

```diff
@@ -363,36 +443,112 @@ impl Client {
         }
     }
 
+    /// Gets whether the topology supports transactions. If it has yet to be determined if the
+    /// topology supports transactions, this method will perform a server selection that will force
+    /// that determination to be made.
+    pub(crate) async fn transaction_support_status(&self) -> Result<TransactionSupportStatus> {
+        let initial_status = self.inner.topology.transaction_support_status().await;
+
+        // Need to guarantee that we're connected to at least one server that can determine if
+        // sessions are supported or not.
+        match initial_status {
+            TransactionSupportStatus::Undetermined => {
+                let criteria = SelectionCriteria::Predicate(Arc::new(move |server_info| {
+                    server_info.server_type().is_data_bearing()
+                }));
+                let _: SelectedServer = self.select_server(Some(&criteria)).await?;
+                Ok(self.inner.topology.transaction_support_status().await)
+            }
+            _ => Ok(initial_status),
+        }
+    }
+
     /// Returns the retryability level for the execution of this operation.
     async fn get_retryability<T: Operation>(
         &self,
         conn: &Connection,
         op: &T,
+        session: &Option<&mut ClientSession>,
     ) -> Result<Retryability> {
-        match op.retryability() {
-            Retryability::Read if self.inner.options.retry_reads != Some(false) => {
-                Ok(Retryability::Read)
-            }
-            Retryability::Write
-                if self.inner.options.retry_writes != Some(false)
-                    && conn.stream_description()?.supports_retryable_writes() =>
-            {
-                Ok(Retryability::Write)
+        if !session
+            .as_ref()
+            .map(|session| session.in_transaction())
+            .unwrap_or(false)
+        {
+            match op.retryability() {
+                Retryability::Read if self.inner.options.retry_reads != Some(false) => {
+                    return Ok(Retryability::Read);
+                }
+                Retryability::Write if conn.stream_description()?.supports_retryable_writes() => {
+                    // commitTransaction and abortTransaction should be retried regardless of the
+                    // value for retry_writes set on the Client
+                    if op.name() == "commitTransaction"
```

Comment:
We can use `CommitTransaction::NAME` and `AbortTransaction::NAME` here (ditto for above).